### DISCUSSION TITLE
ExecutionTime - wall, CPU, both?

### DISCUSSION THREAD

**Discussion by bethac07** [CellProfiler/CellProfiler#4809] [General]
I swear there was a GH issue but I can't find it - just a couple Slack discussions. 

When CellProfiler is running non-built on Mac/Linux and anywhere on Windows, we get as things run a log of the below, which reports CPU time and, since #3091, wall clock execution times. We write execution time to the Image measurements (in ExportToSpreadsheet only, don't even get me started, #4714), but it's CPU time, not wall clock time. We've had a couple discussions in Slack about how users probably don't care about CPU time as much as wall clock time, but haven't actually done anything towards formalizing that. As you can see, in a lot of cases it doesn't matter, but (not shown) sometimes CPU >> wall, indicating "oops, multithreading we didn't expect", and sometimes wall >> CPU, which is often about access to resources.

Last time I brought it up in Slack, Anne why-not-both-memed it, and my response was
>Because already your Image csv has 2x number of modules columns of just like, random reportings (execution errors and timings), 3x seems right out lol

Maybe it's fine though? The average pipeline only has not too many modules, maybe 20-40, so it's not like it's a jillion columns, but for beginner users, any bloat at all is overhwelming. 

I could imagine several scenarios

1. Write CPU time only (current behavior)
2. Write wall clock time only
3. Write both
4. Control which you write via the preferences (in which case, what's our default? Also, do we do that for the error reporting columns too)?
5. Control which you write via module settings in Export (same notes as above)

I think I have a soft preference for 5, with default being "error reporting and wall clock execution on, CPU time off", but I could be argued relatively easily in other directions.

----

>Worker 0: Running module IdentifyPrimaryObjects 14
Worker 0: Tue Aug 15 12:55:53 2023: Image # 239, module IdentifyPrimaryObjects # 14: CPU_time = 0.14 secs, Wall_time = 0.15 secs
Worker 0: Running module MeasureColocalization 15
Worker 3: Tue Aug 15 12:55:08 2023: Image # 240, module RunCellpose # 6: CPU_time = 0.21 secs, Wall_time = 46.27 secs
Worker 3: Running module IdentifySecondaryObjects 8
Worker 3: Tue Aug 15 12:55:55 2023: Image # 240, module IdentifySecondaryObjects # 8: CPU_time = 0.31 secs, Wall_time = 0.69 secs
Worker 3: Running module IdentifyTertiaryObjects 9
Worker 3: Tue Aug 15 12:55:55 2023: Image # 240, module IdentifyTertiaryObjects # 9: CPU_time = 0.10 secs, Wall_time = 0.09 secs
Worker 3: Running module EnhanceOrSuppressFeatures 10
Worker 3: Tue Aug 15 12:55:55 2023: Image # 240, module EnhanceOrSuppressFeatures # 10: CPU_time = 0.07 secs, Wall_time = 0.07 secs
Worker 3: Running module MaskImage 11
Worker 3: Tue Aug 15 12:55:55 2023: Image # 240, module MaskImage # 11: CPU_time = 0.00 secs, Wall_time = 0.00 secs
Worker 3: Running module IdentifyPrimaryObjects 12
Worker 3: Tue Aug 15 12:55:55 2023: Image # 240, module IdentifyPrimaryObjects # 12: CPU_time = 0.21 secs, Wall_time = 0.83 secs
Worker 3: Running module MaskImage 13
Worker 3: Tue Aug 15 12:55:56 2023: Image # 240, module MaskImage # 13: CPU_time = 0.00 secs, Wall_time = 0.01 secs

**Comment by gnodar01** [comment]
Note here: modify #3091 

```
os_times = os.times()
wall_t0 = timeit.default_timer()
cpu_t0 = sum(os_times[:-1])
```

to either

```
os_times = os.times()
wall_t0 = os_times[-1]
cpu_t0 = sum(os_times[:-1])
```

or, as long as `os.times()` doesn't do anything special, then preferably

```
wall_t0 = timeit.time.perf_counter() # aliased by timeit.default_timer()
cpu_t0 = timeit.time.process_time()
```

Mixing and matching is a little confusing.

**Comment by gnodar01** [comment]
> I think I have a soft preference for 5, with default being "error reporting and wall clock execution on, CPU time off"

This would be my preference as well.
